[Skia] Ensure the correct fill sources are applied for each fill paint used

Scony opened the pull request:
#### 459c7fa7db217648a4885d72b971063b30698835
<pre>
[Skia] Ensure the correct fill sources are applied for each fill paint used
<a href="https://bugs.webkit.org/show_bug.cgi?id=271047">https://bugs.webkit.org/show_bug.cgi?id=271047</a>

Reviewed by Carlos Garcia Campos.

Currently, createFillPaint() invoked with clear color may actually pick up a shader instead.
This change separates automatic detection of fill source from createFillPaint() so that it&apos;s not used by mistake.
This change does the same for createStrokePaint().

* Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp:
(WebCore::FontCascade::drawGlyphs):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawRect):
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::drawLine):
(WebCore::GraphicsContextSkia::drawEllipse):
(WebCore::GraphicsContextSkia::fillPath):
(WebCore::GraphicsContextSkia::createFillPaint const):
(WebCore::GraphicsContextSkia::setFillSource const):
(WebCore::GraphicsContextSkia::fillRect):
(WebCore::GraphicsContextSkia::drawDotsForDocumentMarker):
(WebCore::GraphicsContextSkia::clearRect):
(WebCore::GraphicsContextSkia::fillRoundedRectImpl):
(WebCore::GraphicsContextSkia::fillRectWithRoundedHole):
(WebCore::GraphicsContextSkia::drawPattern):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:

Canonical link: <a href="https://commits.webkit.org/276154@main">https://commits.webkit.org/276154@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0bdf9e37bbb083ada0d1ef72abe1432587d3cfb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43929 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22983 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46359 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46574 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40003 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46232 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26887 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20381 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44504 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20017 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37833 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17250 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38920 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1985 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/40103 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39213 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48146 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18933 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/15508 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43072 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20322 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41783 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20528 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6005 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19947 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -274,6 +276,7 @@ void GraphicsContextSkia::drawNativeImageInternal(NativeImage& nativeImage, cons
     SkPaint paint = createFillPaint();
     paint.setBlendMode(toSkiaBlendMode(options.compositeOperator(), options.blendMode()));
     paint.setImageFilter(createDropShadowFilterIfNeeded(ShadowStyle::Outset));
+    setFillSource(paint);
```
I don't think it's needed for images.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -124,7 +124,9 @@ void GraphicsContextSkia::drawRect(const FloatRect& rect, float borderThickness)
 
-    canvas().drawRect(rect, createFillPaint());
+    SkPaint paint = createFillPaint();
+    setFillSource(paint);
```
`setFillSource` being a member of `GraphicsContextSkia` sounds like it's setting something on the context. Maybe we could rename it as `setupFillSource` or something like that.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -289,7 +292,9 @@ void GraphicsContextSkia::drawLine(const FloatPoint& point1, const FloatPoint& p
 
-    canvas().drawLine(SkFloatToScalar(point1.x()), SkFloatToScalar(point1.y()), SkFloatToScalar(point2.x()), SkFloatToScalar(point2.y()), createFillPaint());
+    SkPaint paint;
+    setFillSource(paint);
```
I think only the color makes sense here. Also even if it doesn't solve the problem of dashed borders, we can change the fill to stroke here, since it's clear we want a stroke.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -703,6 +720,7 @@ void GraphicsContextSkia::clearRect(const FloatRect& rect)
 
     auto paint = createFillPaint();
     paint.setBlendMode(SkBlendMode::kClear);
+    setFillSource(paint);
```
I don't think this is needed here, we are clearing.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -842,6 +862,7 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
 
     SkPaint paint = createFillPaint();
     paint.setBlendMode(toSkiaBlendMode(options.compositeOperator(), options.blendMode()));
+    setFillSource(paint);
```
I don't think it's needed for patterns either.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp:
```diff
@@ -54,6 +54,8 @@ void FontCascade::drawGlyphs(GraphicsContext& graphicsContext, const Font& font,
     SkPaint paint = static_cast<GraphicsContextSkia*>(&graphicsContext)->createFillPaint();
     paint.setAntiAlias(font.allowsAntialiasing());
     paint.setImageFilter(static_cast<GraphicsContextSkia*>(&graphicsContext)->createDropShadowFilterIfNeeded(GraphicsContextSkia::ShadowStyle::Outset));
+    static_cast<GraphicsContextSkia*>(&graphicsContext)->setFillSource(paint);
```
I'm not sure about this, I think the fill color is what we always want for text, in that case we can make the helper private.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp:
```diff
@@ -51,9 +51,12 @@ void FontCascade::drawGlyphs(GraphicsContext& graphicsContext, const Font& font,
     auto blob = builder.make();
     auto* canvas = graphicsContext.platformContext();
-    SkPaint paint = static_cast<GraphicsContextSkia*>(&graphicsContext)->createFillPaint();
+    GraphicsContextSkia* skiaGraphicsContext = static_cast<GraphicsContextSkia*>(&graphicsContext);
```
```suggestion
    auto* skiaGraphicsContext = static_cast<GraphicsContextSkia*>(&graphicsContext);
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -289,7 +292,10 @@ void GraphicsContextSkia::drawLine(const FloatPoint& point1, const FloatPoint& p
-    canvas().drawLine(SkFloatToScalar(point1.x()), SkFloatToScalar(point1.y()), SkFloatToScalar(point2.x()), SkFloatToScalar(point2.y()), createFillPaint());
+    // FIXME: Use stroke style to handle dashed/dotted lines, see https://bugs.webkit.org/show_bug.cgi?id=271044
+    SkPaint paint = createFillPaint();
+    paint.setColor(SkColor(this->fillColor().colorWithAlphaMultipliedBy(this->state().alpha())));
```
I guess you can remove `this->` in both cases

Scony committed:
[Skia] Ensure the correct fill sources are applied for each fill paint used
https://bugs.webkit.org/show_bug.cgi?id=271047

Reviewed by Carlos Garcia Campos.

Currently, createFillPaint() invoked with clear color may actually pick up a shader instead.
This change separates automatic detection of fill source from createFillPaint() so that it's not used by mistake.
This change does the same for createStrokePaint().

* Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp:
(WebCore::FontCascade::drawGlyphs):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawRect):
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::drawLine):
(WebCore::GraphicsContextSkia::drawEllipse):
(WebCore::GraphicsContextSkia::fillPath):
(WebCore::GraphicsContextSkia::createFillPaint const):
(WebCore::GraphicsContextSkia::setFillSource const):
(WebCore::GraphicsContextSkia::fillRect):
(WebCore::GraphicsContextSkia::drawDotsForDocumentMarker):
(WebCore::GraphicsContextSkia::clearRect):
(WebCore::GraphicsContextSkia::fillRoundedRectImpl):
(WebCore::GraphicsContextSkia::fillRectWithRoundedHole):
(WebCore::GraphicsContextSkia::drawPattern):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:

Canonical link: https://commits.webkit.org/276154@main
